azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```diff
@@ -58,18 +65,126 @@ public synchronized Connection getConnection() throws SQLException {
          * Every Attempt generates a thread-hanging-time, about 75 seconds, which is hard coded, and can not be changed.
          */
         connection = createDataSource().getConnection();
-        if (connection != null) {
+
+        /**
+         * If connection is null or connection is read only, retry to find available connection.
+         * When DB fails over from master to slave, master is set to read-only mode. We must keep
+         * finding correct data source and sql connection.
+         */
+        if (connection == null || isReadOnly(connection)) {
+          throw new SQLException("Failed to find DB connection Or connection is read only. ");
+        } else {
           return connection;
         }
       } catch (final SQLException ex) {
-        logger.error(
-            "Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt,
-            ex);
-      } finally {
+
+        logger.error( "Failed to find write-enabled DB connection. Wait 1 minutes and retry."
+            + " No.Attempt = " + retryAttempt, ex);
+        /**
+         * When database is completed down, DB connection fails to be fetched immediately. So we need
+         * to hang 1 minute for retry.
+         */
+        sleep(1000L * 60);
```

Comment:
for last retry should we skip the hanging to save some time?

Reply:
also if it runs out of retries due to db is read-only, who is responsible for closing the connection?

Reply:
Based on the current design, we will never run out of retries due to retry maximum limit is extremely high. We can start from this, and think about what we can do if retry number reaches a threshold.

Reply:
I think it would be better to fix it now than later because connection leak is hard to find out later.